### PR TITLE
Simple TTF fallback with additional fixes

### PR DESCRIPTION
**Pull request by SmilyOrg**
Added Font Fallback example
The simple font fallback is activated if you try to draw some text without first loading and setting a valid text format type face. At that point it prints a warning about it and tries to load the per-platform predefined default font if available. If none was found or if the platform isn't implemented, it prints another warning about that and doesn't set any type face (resulting in missing text).

Fixed text format state in nanovg (you could fairly easily break it with unrelated calls before)
Fixed fontstash / nanovg custom alloc implementation

### CODE REVIEW

File: docs/examples/FontFallbackExample/loom.config [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "sdk_version": "dev",
```
Comment (bengarney) [inline docs/examples/FontFallbackExample/loom.config]: Should be latest.

File: loom/common/config/applicationConfig.cpp [hunk 2]
```diff
@@ -125,7 +125,6 @@ static int _jsonParseInt(const char *key, json_t *value)
 
 const utString& LoomApplicationConfig::displayOrientation()
 {
-    lmLog(gLoomApplicationConfigLogGroup, "displayOrientation %s", _displayOrientation.c_str());
```
Comment (bengarney) [inline loom/common/config/applicationConfig.cpp]: Thanks.

File: docs/examples/FontFallbackExample/src/FontFallbackExample.ls [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+{
+    import loom.Application;
+    import loom2d.display.Graphics;
+    import loom2d.display.Quad;
+    import loom2d.display.Shape;
+    import loom2d.display.StageScaleMode;
+    import loom2d.display.SVG;
+    import loom2d.display.TextFormat;
+
+    public class FontFallbackExample extends Application
```
Comment (bengarney) [inline docs/examples/FontFallbackExample/src/FontFallbackExample.ls]: Do we need our own fallback example for this purpose? Should we reuse the NanoVG test?

File: loom/graphics/gfxVectorRenderer.cpp [hunk 4]
```diff
@@ -344,20 +283,143 @@ void VectorRenderer::arc(float x, float y, float radius, float angleFrom, float
 	nvgArc(nvg, x, y, radius, angleFrom, angleTo, direction);
 }
 
+static void readFontFile(const char *path, void* &mem, long &size)
+{
+    FILE *file = fopen(path, "rb");
```
Comment (bengarney) [inline loom/graphics/gfxVectorRenderer.cpp]: This should use platform_mapFile I think/

File: loom/graphics/gfxVectorRenderer.cpp [hunk 5]
```diff
@@ -491,6 +549,17 @@ void VectorTextFormat::load(utString fontName, utString filePath) {
     loom_asset_unlock(filePath.c_str());
 }
 
+void VectorTextFormat::merge(VectorTextFormat* source) {
+	if (source->font.size() > 0) {
+		font = source->font;
+		fontId = VectorTextFormat::FONT_UNDEFINED;
+	}
+	if (source->color >= 0) color = source->color;
+	if (!isnan(source->size)) size = source->size;
```
Comment (bengarney) [inline loom/graphics/gfxVectorRenderer.cpp]: Is isnan cross platform safe?

File: docs/examples/FontFallbackExample/loom.config (hunk 1 above)
Comment (SmilyOrg) [inline docs/examples/FontFallbackExample/loom.config]: Right.

**Comment by bengarney** [comment]
Nitted! Looks good.

File: docs/examples/FontFallbackExample/src/FontFallbackExample.ls (hunk 3 above)
Comment (SmilyOrg) [inline docs/examples/FontFallbackExample/src/FontFallbackExample.ls]: I think it'll be useful when we add additional fallback features (per-character and stuff) and custom system fonts. Might bloat NanoVG a bit too much, but maybe it'd be fine?

File: loom/graphics/gfxVectorRenderer.cpp (hunk 5 above)
Comment (SmilyOrg) [inline loom/graphics/gfxVectorRenderer.cpp]: Yes, we had similar code before, just in a different place.

File: loom/graphics/gfxVectorRenderer.cpp (hunk 4 above)
Comment (SmilyOrg) [inline loom/graphics/gfxVectorRenderer.cpp]: Ah, thought there might be something like that, using it!

**Comment by SmilyOrg** [comment]
Fixed some.

**Comment by SmilyOrg** [comment]
Okay, I'm at least unmapping on error now and I've disabled ownership of nanovg over the memory. Previously nanovg/fontstash would lmFree the memory when it was no longer needed, however now with the mapping API, this doesn't seem possible anymore without copying the memory or changing platform_mapFile?

**Comment by bengarney** [comment]
Please check live reload doesn't make the TTF file re-map.

**Comment by SmilyOrg** [comment]
Ok, this should be fine now.

**Comment by SmilyOrg** [comment]
And of course 10 different builds fail just as I say that.

**Comment by bengarney** [comment]
OK, landing!

**Pull request merged**